### DISCUSSION TITLE
using actions to destroy a record

### DISCUSSION THREAD

**Discussion by vitaly107** [avo-hq/avo#412] [Help] [answered]
I would like to use actions to delete a record and I currently have

```ruby
  def handle(models: :page, fields:)
    models.each do |model|
      if model.published
        fail 'You cannot delete a page that is currently published'
      else
        model.destroy
      end
      succeed 'The record has been deleted'
      redirect_to '/home'
    end
  end
``` 

However this does not work and the object is not deleted/destroyed as the form sends POST request.

How can I send DELETE and make this work? Many thanks in advance.

**Comment by adrianthedev** [comment]
`handle` is a regular ruby method that is ran when the action is used. So you don't have to add the `:page` default param to the `models:` attribute.

```diff
+  def handle(models:, fields:)
-  def handle(models: :page, fields:)
    models.each do |model|
      if model.published
        fail 'You cannot delete a page that is currently published'
      else
        model.destroy
      end
      succeed 'The record has been deleted'
      redirect_to '/home'
    end
  end
```

**Reply by vitaly107** [reply]
I amended as above but the record (model) is still not deleted. any clue why?

**Reply by adrianthedev** [reply]
Please troubleshoot and make sure that the code there gets executed. 

```ruby
  def handle(models:, fields:)
	abort models.inspect
    models.each do |model|
      if model.published
        fail 'You cannot delete a page that is currently published'
      else
        model.destroy
      end
      succeed 'The record has been deleted'
      redirect_to '/home'
    end
  end
```
Add that `abort` and see if the right models get dumped.

**Reply by vitaly107** [reply]
I am getting this in the console and the record still stays:

Started POST "/uk/bj-admin/resources/pages/actions/delete_record" for 127.0.0.1 at 2021-05-27 14:48:13 +0100
Processing by Avo::ActionsController#handle as JS
  Parameters: {"authenticity_token"=>"xxxxxx", "fields"=>{"resource_ids"=>"7"}, "button"=>"", "resource_name"=>"pages", "action_id"=>"delete_record"}
Unpermitted parameters: :authenticity_token, :button
  Page Load (0.4ms)  SELECT "pages".* FROM "pages" WHERE "pages"."id" = $1 LIMIT $2  [["id", 7], ["LIMIT", 1]]
Unpermitted parameters: :authenticity_token, :button
Redirected to http://localhost:3016/uk/bj-admin/resources/pages

**Reply by vitaly107** [reply]
I know form uses POST by default but should not it be DELETE here?

**Reply by vitaly107** [reply]
do I need to add destroy action in avo controller or anything else there? it is empty for now.

**Reply by adrianthedev** [reply]
Did you try the snippet above with the `abort` statement? Did that code get run?

**Reply by adrianthedev** [reply]
Are you still using the `friendly_ids` gem? That might be the issue. The records can't be found because they don't use the `.friendly` query param.

**Reply by vitaly107** [reply]
snippet above with abort does not do anything different from the original version, there was no stopping of execution and teh console output was the same as above

**Reply by vitaly107** [reply]
yes, frinedly ids is still there but the standard way of deletion using your basket button works fine. I cannot see any record nnot found error either. all we are trying to achieve is to make this deletion functionality more prominent and user friendly utilising big button that is provided by actions functionality.

**Reply by adrianthedev** [reply]
The `abort` statement should be reached and halt execution.

**Reply by vitaly107** [reply]
I have this now

**Comment by vitaly107** [comment]
`Started POST "/uk/bj-admin/resources/pages/actions/delete_record" for 127.0.0.1 at 2021-05-27 15:07:36 +0100
Processing by Avo::ActionsController#handle as JS
  Parameters: {"authenticity_token"=>"xxxxx", "fields"=>{"resource_ids"=>"9"}, "button"=>"", "resource_name"=>"pages", "action_id"=>"delete_record"}
Unpermitted parameters: :authenticity_token, :button
  Page Load (0.3ms)  SELECT "pages".* FROM "pages" WHERE "pages"."id" = $1 LIMIT $2  [["id", 9], ["LIMIT", 1]]
Unpermitted parameters: :authenticity_token, :button
Completed 500 Internal Server Error in 10ms (ActiveRecord: 0.3ms | Allocations: 3824)


Not notifying due to notify_release_stages :["staging", "production"]
[#<Page id: 9, name: "Our Team", meta_title: "", meta_description: "", content: "<div>Our team is nice.</div>", created_at: "2021-05-27 14:05:08", updated_at: "2021-05-27 14:05:08", slug: "our-team", published: nil, ancestry: "4">]

SystemExit - [#<Page id: 9, name: "Our Team", meta_title: "", meta_description: "", content: "<div>Our team is nice.</div>", created_at: "2021-05-27 14:05:08", updated_at: "2021-05-27 14:05:08", slug: "our-team", published: nil, ancestry: "4">]:
  app/avo/actions/delete_record.rb:7:in `handle'`

**Reply by adrianthedev** [reply]
So the `abort` works. As I see that record is not published. Maybe that's why it wasn't deleting it?

**Comment by vitaly107** [comment, marked as answer]
I got it working now. Restarted and tested on published page. This can be closed now. Thank you.

**Reply by adrianthedev** [reply]
👍
